client/src/components/routes/privateRoute/index.tsx

```diff
@@ -1,22 +1,83 @@
-import * as React from 'react';
+import React, { useCallback, useEffect, useState } from 'react';
+import { useMutation } from 'react-apollo';
+import { useSelector, useDispatch } from 'react-redux';
 import { Route } from 'react-router';
-import { useQuery } from '@apollo/react-hooks';
 import { Redirect } from 'react-router-dom';
+import axios from 'axios';
 import { Spinner } from '@blueprintjs/core';
+import queryString from 'query-string';
 import { loader } from 'graphql.macro';
-// Graphql
-const meQuery = loader('src/graphql/queries/me.graphql');
+import jwt_decode from 'jwt-decode';
+
+import { fetchTokens, getAccessToken, removeToken } from 'oauth/tokenManager';
+import { onError } from 'services/apollo';
+import { login as loginAction } from 'services/user/actions';
+import { IReduxStore } from 'types';
+import {
+  STATE_STORAGE_KEY,
+  ACCESS_TOKEN_STORAGE_KEY,
+  ID_TOKEN_STORAGE_KEY
+} from '../../../constants';
+
+const mUpsertUser = loader('src/graphql/mutations/upsertUser.graphql');
 
 const PrivateRoute = ({ component: Component, ...rest }: any) => {
-  const { loading, error } = useQuery(meQuery, {
-    fetchPolicy: 'network-only'
+  const dispatch = useDispatch();
+
+  const toaster = useSelector((state: IReduxStore) => state.toaster);
+
+  const onUpsertCompleted = (data: any) => {
+    // Put the user info in redux
+    dispatch(loginAction(data.upsertUser));
+  };
+
+  const [upsertUser] = useMutation(mUpsertUser, {
+    onCompleted: onUpsertCompleted,
+    onError: onError(toaster)
   });
 
-  if (loading) {
-    return <Route component={Spinner} />;
-  }
+  const [token, setToken] = useState(getAccessToken());
+  const params = queryString.parse(window.location.search);
+
+  const storedState = localStorage.getItem(STATE_STORAGE_KEY);
+  const stateMatch =
+    'code' in params && 'state' in params && params.state === storedState;
+
+  const setLoggedInUser = useCallback(async () => {
+    await fetchTokens();
+    const accessToken = localStorage.getItem(ACCESS_TOKEN_STORAGE_KEY);
+    const idToken = localStorage.getItem(ID_TOKEN_STORAGE_KEY);
+    setToken(accessToken);
 
-  if (error) {
+    // Set axios interceptor
+    // TODO put all that's below somewhere else?
+    axios.interceptors.request.use(config => {
+      config.headers.Authorization = `Bearer ${accessToken}`;
+      return config;
+    });
+
+    // Get user info
+    const decodedIdToken: any = jwt_decode(idToken!);
+
+    // Upsert user from IdP into Pyrog DB
+    upsertUser({
```

Comment:
is it the client's responsibility to do that ? I find it safer if it's the serve that updates the user info based on the ID token rather than the client telling pyrog-server to update the user with potentialy arbitrary data

Reply:
I was thinking the same thing. Do you think I should totally remove the upsertUser mutation and upsert at each token introspection in the server (in the getUser function in utils.ts)?

Reply:
I left a comment about that, I think